scripts/bazel_diff.py

```diff
@@ -25,145 +27,186 @@ def log(s: str) -> None:
-    ]
-    p = subprocess.run(
-        args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, encoding="utf-8"
+    return quiet_run_output(
+        [
+            str(bazel_diff),
+            "get-impacted-targets",
+            f"-sh={baseline_hashes}",
+            f"-fh={current_hashes}",
+        ]
     )
-    if p.returncode != 0:
-        print(p.stderr)
-        exit(f"Bazel diff returned {p.returncode}")
-    return p.stdout
 
 
-def filter_targets(bazel: str, targets: str) -> str:
-    args = [
-        bazel,
-        "query",
-        (
+def filter_targets(bazel: Path, targets: str) -> str:
+    with tempfile.NamedTemporaryFile(mode="w+") as tmp:
+        tmp.write(
             f"let t = set({targets}) in "
-            f"kind(rule, $t) except attr(tags, manual, $t)"
-        ),
-    ]
+            "kind(rule, $t) except attr(tags, manual, $t)\n"
+        )
+        tmp.seek(0)
+        tmp_head = "".join(line for (line, _) in zip(tmp, range(10)))
+        if tmp.read(1) != "":
+            tmp_head += "...\n"
+        log(f"Bazel query file's first 10 lines:\n{tmp_head}---")
+        return quiet_run_output(
+            [
+                str(bazel),
+                "query",
+                f"--query_file={tmp.name}",
+            ]
+        )
+
+
+def git_is_dirty() -> bool:
     p = subprocess.run(
-        args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, encoding="utf-8"
+        ["git", "status", "--porcelain", "--untracked-files=no"],
+        stdout=subprocess.PIPE,
+        check=True,
+        encoding="utf-8",
     )
-    if p.returncode != 0:
-        print(p.stderr)
-        exit(f"Bazel run returned {p.returncode}")
-    return p.stdout
+    return len(p.stdout) > 0
 
 
-def git_checkout(commit: str) -> None:
-    subprocess.run(
-        [
-            "git",
-            "checkout",
-            "--quiet",
-            commit,
-        ],
-        check=True,
+def git_current_head() -> str:
+    # Try to get and preserve symbolic-ref if HEAD point at one.
+    p = subprocess.run(
+        ["git", "symbolic-ref", "--quiet", "--short", "HEAD"],
         encoding="utf-8",
+        stdout=subprocess.PIPE,
     )
+    if p.returncode == 0:
+        return p.stdout.strip()
+
+    # Otherwise, just extract the commit.
+    return subprocess.run(
+        ["git", "rev-parse", "--short", "HEAD"],
+        stdout=subprocess.PIPE,
+        check=True,
+        encoding="utf-8",
+    ).stdout.strip()
+
+
+def git_checkout(commit: str) -> None:
+    subprocess.run(["git", "checkout", "--quiet", commit], check=True)
 
 
 def git_diff(baseline: str, current: str) -> None:
-    subprocess.run(
-        [
-            "git",
-            "diff",
-            "--stat",
-            f"{baseline}..{current}",
-        ],
+    p = subprocess.run(
+        ["git", "diff", "--stat", f"{baseline}..{current}"],
+        stdout=subprocess.PIPE,
         check=True,
         encoding="utf-8",
     )
+    log(p.stdout)
```

Comment:
You could do: `stdout=sys.stderr.buffer`, removing the `stdout` and `encoding`. Note that would have the effect of printing output even on errors, when the current approach would suppress.

Reply:
Brilliant, that's actually what I was trying to achieve, done.